src/style-spec/function/compile.js

```diff
@@ -0,0 +1,122 @@
+// @flow
+
+const assert = require('assert');
+module.exports = compileExpression;
+
+const {
+    ParsingContext
+} = require('./expression');
+const parseExpression = require('./parse_expression');
+const { CompoundExpression } = require('./compound_expression');
+const definitions = require('./definitions');
+const evaluationContext = require('./evaluation_context');
+
+import type { Type } from './types.js';
+import type { Expression, ParsingError } from './expression.js';
+
+type CompileErrors = {|
+    result: 'error',
+    errors: Array<ParsingError>
+|}
+
+type CompiledExpression = {|
+    result: 'success',
+    function: Function,
+    functionSource: string,
+    isFeatureConstant: boolean,
+    isZoomConstant: boolean,
+    expression: Expression
+|}
+
+/**
+ *
+ * Given a style function expression object, returns:
+ * ```
+ * {
+ *   result: 'success',
+ *   isFeatureConstant: boolean,
+ *   isZoomConstant: boolean,
+ *   function: Function
+ * }
+ * ```
+ * or else
+ *
+ * ```
+ * {
+ *   result: 'error',
+ *   errors: Array<CompileError>
+ * }
+ * ```
+ *
+ * @private
+ */
+function compileExpression(
+    expr: mixed,
+    expectedType?: Type
+): CompiledExpression | CompileErrors {
+    const context = new ParsingContext(definitions, [], expectedType || null);
+    const parsed = parseExpression(expr, context);
+    if (!parsed) {
+        assert(context.errors.length > 0);
+        return {
+            result: 'error',
+            errors: context.errors
+        };
+    }
+
+    const compiled = parsed.compile();
+    if (typeof compiled === 'string') {
+        const fn = new Function('$globalProperties', '$feature', `
+$globalProperties = $globalProperties || {};
+var $props = $feature && $feature.properties || {};
+var $this = this;
+return this.unwrap(${compiled})
+`);
+
+        return {
+            result: 'success',
+            function: fn.bind(evaluationContext()),
```

Comment:
might be worth trying a custom `bind` or an arrow fn — calls of natively binded functions are expensive in many browsers